### PR TITLE
[SPARK-40003][PYTHON][SQL] Add 'median' to functions

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
expose function `median` in `functions`

### Why are the changes needed?
to support `meidan` in PySpark

### Does this PR introduce _any_ user-facing change?
yes, new API

### How was this patch tested?
doctest

### CODE REVIEW

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -305,6 +305,40 @@ def mean(col: "ColumnOrName") -> Column:
...
+    ...     schema=("course", "year", "earnings"))
+    >>> df.groupby("course").agg(median("earnings")).show()
+    +------+----------------+
+    |course|median(earnings)|
+    +------+----------------+
+    |  Java|         22000.0|
+    |dotNET|         10000.0|
+    +------+----------------+
+    """
+    return _invoke_function_over_columns("median", col)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:339]: Can we use `percentile_approx` instead? Percentile implementation is actually super slow.

Comment (zhengruifeng) [reply python/pyspark/sql/functions.py:339]: What about adding `Note` here? It maybe better to keep in line with the scala side.

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:339]: I think the existing code base use percentile_approx more often:

```
pyspark/pandas/frame.py:            # to `[Column<'percentile_approx(A, 0.2, 10000)'>,
pyspark/pandas/frame.py:            # Column<'percentile_approx(B, 0.2, 10000)'>,
pyspark/pandas/frame.py:            # Column<'percentile_approx(A, 0.5, 10000)'>,
pyspark/pandas/frame.py:            # Column<'percentile_approx(B, 0.5, 10000)'>]`
pyspark/pandas/frame.py:                    map(F.percentile_approx, column_names, [percentile] * column_length)
pyspark/pandas/frame.py:                return F.percentile_approx(spark_column.cast(DoubleType()), qq, accuracy)
pyspark/pandas/generic.py:                return F.percentile_approx(spark_column.cast(DoubleType()), 0.5, accuracy)
pyspark/pandas/groupby.py:                            "percentile_approx(`{0}`, array(0.25, 0.5, 0.75)) as `{1}`".format(
pyspark/pandas/groupby.py:            return F.percentile_approx(col, 0.5, accuracy)
pyspark/pandas/plot/core.py:                F.percentile_approx(
pyspark/pandas/series.py:                    return F.percentile_approx(spark_column.cast(DoubleType()), q_float, accuracy)
```

We could switch to `percentile_approx`, and add a note here. In the test, we can leverage tolerance or control the elements to match the results w/ pandas's

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:339]: oh!

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:339]: Sorry, I thought this is pandas API on Spark

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:339]: my bad. LGTM

**Review by HyukjinKwon — APPROVED** [review]
LGTM. let's add it into the Python reference API docs

**Comment by zhengruifeng** [comment]
merged to master, thanks @HyukjinKwon for reviews

**Pull request closed without merging**